From df12b9a8e87d9103a2a39c9353881cc9b4adfbba Mon Sep 17 00:00:00 2001
From: Antoine Cotten <hello@acotten.com>
Date: Thu, 14 Apr 2022 17:46:42 +0200
Subject: [PATCH] Set netdev->dev_addr through dev_addr_mod

Since Linux 5.17 netdev->dev_addr is const and must be changed through
dev_addr_mod.

See also: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=adeef3e32146a8d2a73c399dc6f5d76a449131b1

Backported from RPM Fusion:
https://pkgs.rpmfusion.org/cgit/nonfree/wl-kmod.git/commit/?id=c89e5d9808189084cf2f48b9040b4b00939bd180

See also alternative patch by Joan Bruguera, using eth_hw_addr_set:
https://gist.github.com/joanbm/37fd0590f30b41e7e89ade6e242ca16e
https://github.com/torvalds/linux/commit/48eab831ae8b9f7002a533fa4235eed63ea1f1a3
---
 src/wl/sys/wl_linux.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/wl/sys/wl_linux.c b/src/wl/sys/wl_linux.c
index dee4b9e..9de2cc5 100644
--- a/src/wl/sys/wl_linux.c
+++ b/src/wl/sys/wl_linux.c
@@ -640,7 +640,11 @@ wl_attach(uint16 vendor, uint16 device, ulong regs,
 			WL_ERROR(("wl%d: Error setting MAC ADDRESS\n", unit));
 	}
 #endif 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
 	bcopy(&wl->pub->cur_etheraddr, dev->dev_addr, ETHER_ADDR_LEN);
+#else
+	dev_addr_mod(dev, 0, &wl->pub->cur_etheraddr, ETHER_ADDR_LEN);
+#endif
 
 	online_cpus = 1;
 
@@ -1856,7 +1860,11 @@ wl_set_mac_address(struct net_device *dev, void *addr)
 
 	WL_LOCK(wl);
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
 	bcopy(sa->sa_data, dev->dev_addr, ETHER_ADDR_LEN);
+#else
+	dev_addr_mod(dev, 0, sa->sa_data, ETHER_ADDR_LEN);
+#endif
 	err = wlc_iovar_op(wl->wlc, "cur_etheraddr", NULL, 0, sa->sa_data, ETHER_ADDR_LEN,
 		IOV_SET, (WL_DEV_IF(dev))->wlcif);
 	WL_UNLOCK(wl);
@@ -3040,7 +3048,11 @@ _wl_add_monitor_if(wl_task_t *task)
 	else
 		dev->type = ARPHRD_IEEE80211_RADIOTAP;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
 	bcopy(wl->dev->dev_addr, dev->dev_addr, ETHER_ADDR_LEN);
+#else
+	dev_addr_mod(dev, 0, wl->dev->dev_addr, ETHER_ADDR_LEN);
+#endif
 
 #if defined(WL_USE_NETDEV_OPS)
 	dev->netdev_ops = &wl_netdev_monitor_ops;
